Azure Cosmos DB for MongoDB extension

aayush3011 opened the pull request:
## Motivation and Context (Why the change? What's the scenario?)
I have added Azure Cosmos DB Mongo vCore as Memory to be used by customers using Kernel Memory and Cosmos DB for their LLM applications.

[https://learn.microsoft.com/en-us/azure/cosmos-db/mongodb/vcore/vector-search](url)

dluc commented on extensions/AzureCosmosDBMongoVCore/AzureCosmosDBMongoVCore/AzureCosmosDBMongoVCoreConfig.cs:
```diff
@@ -0,0 +1,70 @@
+{
+    /// <summary>
+    /// Get more details about Azure Cosmos Mongo vCore https://learn.microsoft.com/en-us/azure/cosmos-db/mongodb/vcore/vector-search
+    /// </summary>
```
this comment seems out of place, maybe a property is missing, or it should be above the class declaration?

aayush3011 replied:
Fixed.

dluc commented on extensions/AzureCosmosDBMongoVCore/AzureCosmosDBMongoVCore/AzureCosmosDBMongoVCoreConfig.cs:
```diff
@@ -0,0 +1,70 @@
+    /// </summary>
+
+    /// <summary>
+    /// Connection string required to connect to Azure Cosmos Mongo vCore
```
pls add an example in the description, or a link to docs to find out what this looks like

aayush3011 replied:
Added a reference link.

dluc commented on extensions/AzureCosmosDBMongoVCore/AzureCosmosDBMongoVCore/AzureCosmosDBMongoVCoreConfig.cs:
```diff
@@ -0,0 +1,70 @@
+    ///         - vector-hnsw: available as a preview feature only,
+    ///                        to enable visit https://learn.microsoft.com/en-us/azure/azure-resource-manager/management/preview-features
+    /// </summary>
+    public string Kind { get; set; }
```
any chance this could be an enum?

aayush3011 replied:
Created an enum for this.

dluc commented on extensions/AzureCosmosDBMongoVCore/AzureCosmosDBMongoVCore/AzureCosmosDBMongoVCoreConfig.cs:
```diff
@@ -0,0 +1,70 @@
+
+namespace Microsoft.KernelMemory;
+
+public class AzureCosmosDBMongoVCoreConfig
```
could set a default value for all the props?

dluc commented on extensions/AzureCosmosDBMongoVCore/AzureCosmosDBMongoVCore/AzureCosmosDBMongoVCoreConfig.cs:
```diff
@@ -0,0 +1,70 @@
+    ///         - CosmosDBSimilarityType.L2 (Euclidean distance), and
+    ///         - CosmosDBSimilarityType.IP (inner product).
+    /// </summary>
+    public string Similarity { get; set; }
```
ditto, enum?

aayush3011 replied:
Created an enum for this.

dluc commented on extensions/AzureCosmosDBMongoVCore/AzureCosmosDBMongoVCore/AzureCosmosDBMongoVCoreConfig.cs:
```diff
@@ -0,0 +1,70 @@
+    /// <summary>
+    /// Dimensions: Number of dimensions for vector similarity. The maximum number of supported dimensions is 2000
+    /// </summary>
+    public int Dimensions { get; set; }
```
KM usually calculates this at runtime, and uses the value detected to create indexes of the right size.

aayush3011 replied:
Removed dimensions from the config.

dluc commented on extensions/AzureCosmosDBMongoVCore/AzureCosmosDBMongoVCore/AzureCosmosDBMongoVCoreConfig.cs:
```diff
@@ -0,0 +1,70 @@
+    /// NumberOfConnections: The max number of connections per layer (16 by default, minimum value is 2, maximum value is 
+    /// 100). Higher m is suitable for datasets with high dimensionality and/or high accuracy requirements.
+    /// </summary>
+    public int NumberOfConnections { get; set; }
```
if higher is better, why would one choose 2 instead of 100?

aayush3011 replied:
If the dimensionality of the dataset is low, then you won't need a higher value of "m(NumberOfConnections)". 

Smaller values of "m" tend to be better for lower recalls or low-dimensional data while higher values of"m" are suited better for high recalls or high-dimensional data.

dluc commented on extensions/AzureCosmosDBMongoVCore/AzureCosmosDBMongoVCore/AzureCosmosDBMongoVCoreMemory.cs:
```diff
@@ -0,0 +1,300 @@
+/// </summary>
+
+public class AzureCosmosDBMongoVCoreMemory : IMemoryDb
+{
```
here and other classes: could you sort methods by visibility, ctor, then public, then private at the bottom?

sivethe commented on extensions/AzureCosmosDBMongoVCore/AzureCosmosDBMongoVCore.FunctionalTests/appsettings.json:
```diff
@@ -0,0 +1,73 @@
+    },
+    "KernelMemory": {
+        "Services": {
+            "AzureCosmosDBMongoVCore": {
```
Can we add a config for AppName and a default appName please?

aayush3011 replied:
Added AppName.

sivethe commented on extensions/AzureCosmosDBMongoVCore/AzureCosmosDBMongoVCore/AzureCosmosDBMongoVCoreMemory.cs:
```diff
@@ -0,0 +1,299 @@
+                throw new AzureCosmosDBMongoVCoreMemoryException("Embedding generator not configured");
+            }
+
+            this._cosmosDBMongoClient = new MongoClient(config.ConnectionString);
```
Create MongoClientSettings first and then the MongoClient so you can pass the appName

e.g.
MongoClientSettings settings = MongoClientSettings.FromConnectionString(config.ConnectionString);
settings.ApplicationName = config.ApplicationName;
this._cosmosDBMongoClient = new MongoClient(settings);

aayush3011 replied:
Added the code creating AppName.

dluc commented on extensions/AzureCosmosDBMongoVCore/AzureCosmosDBMongoVCore.TestApplication/appsettings.json:
```diff
@@ -0,0 +1,68 @@
+    "UseRealEmbeddingGenerator": false,
+    "KernelMemory": {
+        "Services": {
+            "AzureCosmosDBMongoVCore": {
```
I believe this should be renamed to `AzureCosmosDBMongoDBVCore`

aayush3011 replied:
Updated the name

dluc commented on extensions/AzureCosmosDBMongoVCore/AzureCosmosDBMongoVCore/AzureCosmosDBMongoVCore.csproj:
```diff
@@ -0,0 +1,46 @@
+    <PropertyGroup>
+        <TargetFramework>net6.0</TargetFramework>
+        <RollForward>LatestMajor</RollForward>
+        <AssemblyName>Microsoft.KernelMemory.MemoryDb.AzureCosmosDBMongoVCore</AssemblyName>
```
same here and other places, I think the name is missing a `DB` after Mongo. 

For instance this one would be `Microsoft.KernelMemory.MemoryDb.AzureCosmosDBMongoDBVCore`.

Same for namespace and class names

aayush3011 replied:
Updated the name

dluc commented on extensions/AzureCosmosDBMongoVCore/AzureCosmosDBMongoVCore/AzureCosmosDBMongoVCore.csproj:
```diff
@@ -0,0 +1,46 @@
+    </ItemGroup>
+       
+    <ItemGroup>
+        <PackageVersion Include="MongoDB.Driver" Version="2.24.0" />
```
pls use PackageReference and the centralized versioning

dluc commented on extensions/AzureCosmosDBMongoVCore/AzureCosmosDBMongoVCore/AzureCosmosDBMongoVCore.csproj:
```diff
@@ -0,0 +1,46 @@
+    <PropertyGroup>
+        <IsPackable>true</IsPackable>
+        <PackageId>Microsoft.KernelMemory.MemoryDb.AzureCosmosDBMongoVCore</PackageId>
+        <Product>Azure CosmosDB Mongo vCore connector for Kernel Memory</Product>
```
missing `DB` here too

aayush3011 replied:
Updated the name

dluc commented on extensions/AzureCosmosDBMongoVCore/AzureCosmosDBMongoVCore/AzureCosmosDBMongoVCoreConfig.cs:
```diff
@@ -0,0 +1,116 @@
+namespace Microsoft.KernelMemory;
+
+/// <summary>
+/// Get more details about Azure Cosmos Mongo vCore and these configs https://learn.microsoft.com/en-us/azure/cosmos-db/mongodb/vcore/vector-search
```
MongoDB. I'll leave to you to find all the places :-)

aayush3011 replied:
Updated the name

dluc commented on extensions/AzureCosmosDBMongoVCore/AzureCosmosDBMongoVCore/AzureCosmosDBMongoVCoreConfig.cs:
```diff
@@ -0,0 +1,116 @@
+
+public enum AzureCosmosDBVectorSearchType
+{
+    [EnumMember(Value = "vector-ivf")]
```
given the need to also JSON serialize these values, could we avoid `EnumMember` and `EnumExtensions` and just rely on `[JsonConverter(typeof(JsonStringEnumConverter))]` and serialize the value when needed?

aayush3011 replied:
Updated the code

dluc reviewed (changes requested):
Please add the project to the solution file. I see a lot of build warnings and errors due to code style, could you take a look?

aayush3011 commented:
@dluc I have added the projects to the Solution file, and renamed the AzureCosmosDBMongoDBVCore to AzureCosmosDBMongoDB as decided by the team for the external connectors/integrations.

dluc commented:
I fixed some compilation errors, but there are still several warnings to address:

```
extensions/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB/AzureCosmosDBMongoDBMemory.cs(35,74): warning CS1570: XML comment has badly formed XML -- 'End tag 'parama' does not match the start tag 'param'.' [extensions/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB.csproj]
extensions/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB/AzureCosmosDBMongoDBMemory.cs(122,59): warning CS8425: Async-iterator 'AzureCosmosDBMongoDBMemory.GetSimilarListAsync(string, string, ICollection<MemoryFilter>?, double, int, bool, CancellationToken)' has one or more parameters of type 'CancellationToken' but none of them is decorated with the 'EnumeratorCancellation' attribute, so the cancellation token parameter from the generated 'IAsyncEnumerable<>.GetAsyncEnumerator' will be unconsumed [extensions/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB.csproj]
extensions/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB/AzureCosmosDBMongoDBMemory.cs(163,49): warning CS8425: Async-iterator 'AzureCosmosDBMongoDBMemory.GetListAsync(string, ICollection<MemoryFilter>?, int, bool, CancellationToken)' has one or more parameters of type 'CancellationToken' but none of them is decorated with the 'EnumeratorCancellation' attribute, so the cancellation token parameter from the generated 'IAsyncEnumerable<>.GetAsyncEnumerator' will be unconsumed [extensions/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB.csproj]
extensions/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB/AzureCosmosDBMongoDBConfig.cs(74,12): warning CS8618: Non-nullable property 'ConnectionString' must contain a non-null value when exiting constructor. Consider declaring the property as nullable. [extensions/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB.csproj]
extensions/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB/AzureCosmosDBMongoDBMemoryRecord.cs(46,41): warning CS8604: Possible null reference argument for parameter 'value' in 'void Dictionary<string, object>.Add(string key, object value)'. [extensions/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB.csproj]
extensions/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB/AzureCosmosDBMongoDBMemoryRecord.cs(21,20): warning CS8618: Non-nullable property 'Embedding' must contain a non-null value when exiting constructor. Consider declaring the property as nullable. [extensions/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB.csproj]
extensions/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB/AzureCosmosDBMongoDBMemory.cs(135,35): warning CS8600: Converting null literal or possible null value to non-nullable type. [extensions/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB.csproj]
extensions/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB/AzureCosmosDBMongoDBMemory.cs(163,49): warning CS1998: This async method lacks 'await' operators and will run synchronously. Consider using the 'await' operator to await non-blocking API calls, or 'await Task.Run(...)' to do CPU-bound work on a background thread. [extensions/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB.csproj]
extensions/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB/AzureCosmosDBMongoDBMemoryRecord.cs(21,20): warning CA1819: Properties should not return arrays (https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1819) [extensions/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB.csproj]
extensions/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB/AzureCosmosDBVectorSearchType.cs(9,5): warning CA1707: Remove the underscores from member name Microsoft.KernelMemory.MemoryDb.AzureCosmosDBMongoDB.AzureCosmosDBVectorSearchType.vector_ivf (https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1707) [extensions/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB.csproj]
extensions/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB/AzureCosmosDBVectorSearchType.cs(10,5): warning CA1707: Remove the underscores from member name Microsoft.KernelMemory.MemoryDb.AzureCosmosDBMongoDB.AzureCosmosDBVectorSearchType.vector_hnsw (https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1707) [extensions/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB.csproj]
extensions/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB/AzureCosmosDBMongoDBMemory.cs(73,47): warning CA1725: In member Task AzureCosmosDBMongoDBMemory.CreateIndexAsync(string indexName, int vectorSize, CancellationToken cancellationToken = default(CancellationToken)), change parameter name indexName to index in order to match the identifier as it has been declared in Task IMemoryDb.CreateIndexAsync(string index, int vectorSize, CancellationToken cancellationToken = default(CancellationToken)) (https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1725) [extensions/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB.csproj]
extensions/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB/AzureCosmosDBMongoDBMemory.cs(75,29): warning CA2016: Forward the 'cancellationToken' parameter to the 'ListAsync' method or pass in 'CancellationToken.None' explicitly to indicate intentionally not propagating the token (https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca2016) [extensions/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB.csproj]
extensions/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB/AzureCosmosDBMongoDBMemory.cs(76,14): warning CA2016: Forward the 'cancellationToken' parameter to the 'ToList' method or pass in 'CancellationToken.None' explicitly to indicate intentionally not propagating the token (https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca2016) [extensions/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB.csproj]
extensions/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB/AzureCosmosDBMongoDBMemory.cs(89,19): warning CA2016: Forward the 'cancellationToken' parameter to the 'RunCommandAsync' method or pass in 'CancellationToken.None' explicitly to indicate intentionally not propagating the token (https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca2016) [extensions/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB.csproj]
extensions/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB/AzureCosmosDBMongoDBMemory.cs(105,15): warning CA2016: Forward the 'cancellationToken' parameter to the 'DropOneAsync' method or pass in 'CancellationToken.None' explicitly to indicate intentionally not propagating the token (https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca2016) [extensions/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB.csproj]
extensions/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB/AzureCosmosDBMongoDBMemory.cs(146,34): warning CA2016: Forward the 'cancellationToken' parameter to the 'AggregateAsync' method or pass in 'CancellationToken.None' explicitly to indicate intentionally not propagating the token (https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca2016) [extensions/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB.csproj]
extensions/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB/AzureCosmosDBMongoDBMemory.cs(249,92): warning CA1305: The behavior of 'float.ToString()' could vary based on the current user's locale settings. Replace this call in 'AzureCosmosDBMongoDBMemory.GetVectorIVFSearchPipeline(Embedding, int)' with a call to 'float.ToString(IFormatProvider)'. (https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1305) [extensions/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB.csproj]
extensions/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB/AzureCosmosDBMongoDBMemory.cs(279,92): warning CA1305: The behavior of 'float.ToString()' could vary based on the current user's locale settings. Replace this call in 'AzureCosmosDBMongoDBMemory.GetVectorHNSWSearchPipeline(Embedding, int)' with a call to 'float.ToString(IFormatProvider)'. (https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1305) [extensions/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB.csproj]
extensions/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB/AzureCosmosDBMongoDBMemory.cs(27,35): warning CA1859: Change type of field '_cosmosDBMongoClient' from 'MongoDB.Driver.IMongoClient' to 'MongoDB.Driver.MongoClient' for improved performance (https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1859) [extensions/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB.csproj]
```

I'm also getting a compilation error:

> /usr/local/share/dotnet/sdk/8.0.201/Sdks/NuGet.Build.Tasks.Pack/build/NuGet.Build.Tasks.Pack.targets(221,5): error NU5019: File not found: 'extensions/AzureCosmosDBMongoDB/README.md'. [extensions/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB.csproj]

dluc commented:
Could you please also squash the commits into a single commit? that will make it easier to manage the PR while other PRs get merged. Thanks!

dluc commented on extensions/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB/AzureCosmosDBVectorSearchType.cs:
```diff
@@ -0,0 +1,11 @@
+public enum AzureCosmosDBVectorSearchType
+{
+    vector_ivf,
+    vector_hnsw
```
please rename these to follow the naming convention, something like `VectorIVF` and `VectorHNSW` would be better. If different values are needed for serialization, you can use the serialization attributes. Thanks

aayush3011 replied:
Updated the values.

dluc commented on extensions/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB/AzureCosmosDBSimilarityType.cs:
```diff
@@ -0,0 +1,12 @@
+{
+    COS,
+    IP,
+    L2
```
please extend the names for readability, something like `Cosine`, `InnerProduct` etc would be better. 
If different values are needed for serialization, you can use the serialization attributes. Thanks

aayush3011 replied:
Updated the values.

dluc commented on extensions/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB.TestApplication/appsettings.json:
```diff
@@ -0,0 +1,68 @@
+        "Services": {
+            "AzureCosmosDBMongoDB": {
+                // For other options see <AzureAISearchConfig>.
+                "ConnectionString": "mongodb+srv://akataria3011:Basket24ball@akatarialangchaintesting.mongocluster.cosmos.azure.com/?tls=true&authMechanism=SCRAM-SHA-256&retrywrites=false&maxIdleTimeMS=120000",
```
@aayush3011 looks like you checked in some personal credentials. Please invalidate the deployment ASAP thanks

aayush3011 replied:
Removed

dluc replied:
@aayush3011 double checking, did you also delete the resource in the portal? 

(the secret, even if removed from the PR, is still visible in git history)

dluc reviewed (changes requested):
Build issues to be addressed before testing - thanks

aayush3011 commented:
@dluc I have addressed all your comments, the build should also be passing now. Let me know if something else is required from my side.

Josephrp reviewed (approved):
fantastic effort, thanks for following through

dluc commented on extensions/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB/AzureCosmosDBSimilarityType.cs:
```diff
@@ -0,0 +1,16 @@
+
+public enum AzureCosmosDBSimilarityType
+{
+    [JsonProperty("COS")]
```
can we avoid using Newtonsoft, and rely on .NET core JSON serialization?

dluc commented on extensions/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB/AzureCosmosDBSimilarityType.cs:
```diff
@@ -0,0 +1,16 @@
+    [JsonProperty("IP")]
+    InnerProduct,
+    [JsonProperty("L2")]
+    Eucledian
```
typo

dluc commented on extensions/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB/AzureCosmosDBSimilarityType.cs:
```diff
@@ -0,0 +1,16 @@
+
+namespace Microsoft.KernelMemory.MemoryDb.AzureCosmosDBMongoDB;
+
+public enum AzureCosmosDBSimilarityType
```
nit: please use plural `AzureCosmosDBSimilarityType`

dluc commented on extensions/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB/AzureCosmosDBVectorSearchType.cs:
```diff
@@ -0,0 +1,15 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+// ReSharper disable InconsistentNaming
+using Newtonsoft.Json;
```
ditto, I would avoid introducing Newtonsoft in the project, and use .NET default JSON serializer instead

dluc commented on Directory.Packages.props:
```diff
@@ -21,8 +21,10 @@
+    <PackageVersion Include="MongoDB.Driver" Version="2.24.0" />
     <PackageVersion Include="MongoDB.Driver.GridFS" Version="2.25.0" />
     <PackageVersion Include="Moq" Version="4.20.70" />
+    <PackageVersion Include="Newtonsoft.Json" Version="13.0.3" />
```
ditto, I would avoid introducing Newtonsoft in the project

dluc commented on extensions/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB.FunctionalTests/appsettings.json:
```diff
@@ -0,0 +1,71 @@
+    },
+    "KernelMemory": {
+        "Services": {
+            "AzureCosmosDBMongoDB": {
```
could you add this block also to `service/Service/appsettings.json`, under `Services`? (the list is sorted alphabetically).
Please add a comment for every line.
About the `ConnectionString` please leave it empty or add a template showing what the cs looks like, that will help users using the right value.

dluc commented on extensions/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB.csproj:
```diff
@@ -0,0 +1,42 @@
+    <Import Project="../../../nuget-package.props" />
+
+    <PropertyGroup>
+        <IsPackable>true</IsPackable>
```
```suggestion
        <IsPackable>false</IsPackable>
```
Let's keep this disabled for now, we need to do some testing after merging, and make other changes to docs and DI (no need to make them now)

aayush3011 replied:
Sure, I'll add filters logic in the next iteration of the PR, in a couple of weeks.

dluc commented on extensions/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB/AzureCosmosDBMongoDBMemory.cs:
```diff
@@ -0,0 +1,308 @@
+            throw new AzureCosmosDBMongoDBMemoryException("Embedding generator not configured");
+        }
+
+        MongoClientSettings settings = MongoClientSettings.FromConnectionString(config.ConnectionString);
```
does AzureMongoDb support Managed Identities and `DefaultAzureCredential`?

aayush3011 replied:
I have added MongoClient to be passes as a variable in the AzureCosmosDBMongoDBMemory constructor. Now if the users want to use AzureKeyCredential, they can use it and directly send the MongoClient to the ctor, or use the vanilla version and pass in only the ConnectionString of the MongoDB cluster in the AzureCosmosDBMongoDBConfig.

dluc replied:
I'd suggest to take a look at the other Azure connectors, they all support Managed Identity. The approach you suggest, about injecting a custom client, would require custom code for anyone using Prod deployments. 

For instance if you look at the "infra" folder and the "Deploy to Azure" button, it allows to deploy the solution in Prod with Managed Identity without changing the code, simply by setting "AuthType" to "AzureIdentity" in the config file.

dluc commented on extensions/AzureCosmosDBMongoDB/README.md:
```diff
@@ -0,0 +1,5 @@
+
+Integrated Vector Database in Azure Cosmos DB for MongoDB vCore can be used to seamlessly connect your AI-based applications with your data that's stored in Azure Cosmos DB. This integration can include apps that you built by using Azure OpenAI embeddings. The natively integrated vector database enables you to efficiently store, index, and query high-dimensional vector data that's stored directly ...
+
+Please refer to this link for setting up your Azure CosmosDB MongoDB vCore cluster, and more information on vector search using [Azure CosmosDB MongoDB vCore](https://learn.microsoft.com/en-us/azure/cosmos-db/mongodb/vcore/vector-search).
```
```suggestion
Please refer to this link for setting up your Azure CosmosDB MongoDB vCore cluster, and more information on vector search using [Azure CosmosDB MongoDB vCore](https://learn.microsoft.com/azure/cosmos-db/mongodb/vcore/vector-search).
```

dluc commented on extensions/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB/AzureCosmosDBEnums.cs:
```diff
@@ -0,0 +1,14 @@
+
+namespace Microsoft.KernelMemory.MemoryDb.AzureCosmosDBMongoDB;
+
+public class AzureCosmosDBEnums
```
this class doesn't seem to be used

dluc commented on extensions/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB/AzureCosmosDBMongoDBMemory.cs:
```diff
@@ -0,0 +1,309 @@
+    public async IAsyncEnumerable<(MemoryRecord, double)> GetSimilarListAsync(
+        string index,
+        string text,
+        ICollection<MemoryFilter>? filters = null,
```
`filters` is not implemented. Does it mean the connector is still underdevelopment?

aayush3011 replied:
The filters are in public preview now, I'll add it in the next iteration in a couple weeks. Does this block the PR right now?

dluc commented on extensions/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB/AzureCosmosDBMongoDBMemory.cs:
```diff
@@ -0,0 +1,309 @@
+        [EnumeratorCancellation] CancellationToken cancellationToken = default)
+    {
+        // TODO: Add logic for search with filters, once it is released in Azure Cosmos DB for MongoDB.
+        // For now this method returns an empty list.
```
does it mean the connector is still underdevelopment?

aayush3011 replied:
The filters are in public preview now, I'll add it in the next iteration in a couple of weeks. Does this block the PR right now?

dluc replied:
Without filtering the connector would not be useful, because it would prevent from using security filters

dluc commented on extensions/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB/AzureCosmosDBMongoDBMemory.cs:
```diff
@@ -0,0 +1,320 @@
+    /// <param name="databaseName">Database name for MongoDB</param>
+    /// <param name="collectionName">Collection name for MongoDB</param>
+    /// <param name="log">Application logger</param>
+    public AzureCosmosDBMongoDBMemory(
```
.NET DI is unable to use this ctor and create instances of the class, because it requires `string` parameters to be resolved:
```
ystem.InvalidOperationException
Unable to resolve service for type 'System.String' while attempting to activate 'Microsoft.KernelMemory.MemoryDb.AzureCosmosDBMongoDB.AzureCosmosDBMongoDBMemory'.
```

You should remove the `collectionName` and use the `index` parameter received by each method below, and move the `databaseName` parameter into the config class.

dluc commented on extensions/AzureCosmosDBMongoDB/AzureCosmosDBMongoDB.FunctionalTests/DefaultTests.cs:
```diff
@@ -0,0 +1,70 @@
+
+namespace AzureCosmosDBMongoDB.FunctionalTests;
+
+public class DefaultTests : BaseFunctionalTestCase
```
please run the tests and make sure they are all passing

dluc reviewed (changes requested):
Major things to be addressed:
* Functional tests should pass
* Filtering should be implemented
* Remove Newtonsoft dependency

dluc commented:
@aayush3011 how is it going with functional tests and filters? Did you consider reusing the logic from the MongoDbAtlas connector?

dluc commented:
Looks like the PR got stale, closing it for now. We can reopen it if someone can complete the connector and address the issues identified.

aayush3011 commented:
@dluc any updates on this?

dluc commented:
Closing for now. We'll reconsider once the new SK connector is ready.
